compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt

```diff
@@ -0,0 +1,214 @@
+            for (event in eventSet) {
+                when (event) {
+                    is VMDeathEvent, is VMDisconnectEvent -> {
+                        break@vmLoop
+                    }
+                    is VMStartEvent -> {
+                        val classReq = manager.createClassPrepareRequest()
+                        classReq.addClassFilter(MAIN_CLASS)
+                        classReq.setSuspendPolicy(SUSPEND_ALL)
+                        classReq.enable()
+                        val vmDeathReq = manager.createVMDeathRequest()
+                        vmDeathReq.setSuspendPolicy(SUSPEND_ALL)
+                        vmDeathReq.enable()
+
+                        virtualMachine.resume()
+                    }
+                    is ClassPrepareEvent -> {
+                        val methodEntryReq = manager.createMethodEntryRequest()
+                        methodEntryReq.addClassFilter(MAIN_CLASS)
+                        methodEntryReq.setSuspendPolicy(SUSPEND_NONE)
+                        methodEntryReq.enable()
+
+                        val methodExitReq = manager.createMethodExitRequest()
+                        methodExitReq.addClassFilter(MAIN_CLASS)
+                        methodExitReq.setSuspendPolicy(SUSPEND_NONE)
+                        methodExitReq.enable()
+
+                        val stepReq = manager.createStepRequest(event.thread(), StepRequest.STEP_LINE, StepRequest.STEP_INTO)
+                        stepReq.setSuspendPolicy(SUSPEND_NONE)
+                        stepReq.addClassExclusionFilter("java.*")
+                        stepReq.addClassExclusionFilter("sun.*")
+                        stepReq.enable()
+
+                        virtualMachine.resume()
+                    }
+                    is MethodEntryEvent -> {
+                        if (event.location().method().name() == BOX_METHOD) {
+                            shouldRecordSteps = true
+                        }
+                    }
+                    is StepEvent -> {
+                        if (shouldRecordSteps) {
+                            loggedEvents.add(event)
+                        }
+                    }
+                    is MethodExitEvent -> {
+                        if (event.location().method().name() == BOX_METHOD) {
+                            manager.stepRequests().forEach { it.disable() }
+                            shouldRecordSteps = false
+                        }
+                    }
+                    else -> {
+                        throw IllegalStateException("event not handled: $event")
+                    }
+                }
+            }
+        }
+        val actualLineNumbers = loggedEvents
+            .map { event ->
+                "${(event as LocatableEvent).location().method()}:${event.location().lineNumber()}"
+            }
+        TestCase.assertEquals(expectedLineNumbers, actualLineNumbers.joinToString(" "))
+    }
+
+    private fun isPortOpen(): Boolean {
```

Comment:
You should be able to avoid this if you read stdout from the jvm for the port number.